databuilder/validate_dummy_data.py

```diff
@@ -1,151 +1,159 @@
+import csv
+import dataclasses
+import datetime
+import gzip
 
-import pandas as pd
+from databuilder.functools_utils import singledispatch_on_value
+from databuilder.query_model import get_series_type
+from databuilder.query_utils import get_column_definitions
 
-from .query_engines.query_model_old import ValueFromCategory, ValueFromRow
-from .query_utils import get_column_definitions
 
-SUPPORTED_FILE_FORMATS = ["csv", "csv.gz"]
+class ValidationError(Exception):
+    pass
 
 
-class DummyDataValidationError(Exception):
-    pass
+@dataclasses.dataclass(frozen=True)
+class ColumnSpec:
+    type: type  # noqa: A003
+    nullable: bool = True
 
 
-def validate_dummy_data(dataset_class, dummy_data_file, dataset_file):
-    """Validate that dummy data provided by user matches expected structure and format.
+def validate_file_types_match(dummy_filename, output_filename):
+    if get_file_type(dummy_filename) != get_file_type(output_filename):
+        raise ValidationError(
+            f"Dummy data file does not have the same file extension as the output "
+            f"filename:\n"
+            f"Dummy data file: {dummy_filename}\n"
+            f"Output file: {output_filename}"
+        )
 
-    Raises DummyDataValidationError if dummy data is not valid.
-    """
 
-    validate_file_extension(dataset_file, dummy_data_file)
+def validate_dummy_data_file(dataset_definition, filename):
+    column_definitions = get_column_definitions(dataset_definition)
+    column_specs = get_column_specs(column_definitions)
 
-    df = read_into_dataframe(dummy_data_file)
+    file_type, gzipped = get_file_type(filename)
+    if file_type != ".csv":
+        raise ValidationError(f"Unsupported file type: {file_type}")
 
-    column_definitions = get_column_definitions(dataset_class)
+    open_fn = gzip.open if gzipped else open
+    with open_fn(filename, mode="rt", newline="") as f:
+        validate_csv_against_spec(f, column_specs)
 
-    # Ignore the population definition, since it is not used as a column in the
-    # dataset
-    column_definitions.pop("population", None)
-    # Add in the patient_id, which is always included as a column in the dataset
-    column_definitions["patient_id"] = ValueFromRow(source=None, column="patient_id")
 
-    validate_expected_columns(df, column_definitions)
+def get_file_type(filename):
+    suffixes = filename.suffixes
+    if suffixes and suffixes[-1] == ".gz":
+        gzipped = True
+        suffixes.pop()
+    else:
+        gzipped = False
+    extension = suffixes[-1] if suffixes else ""
+    return extension, gzipped
 
-    validate_column_values(df, column_definitions)
 
+def get_column_specs(column_definitions):
+    # TODO: It may not be universally true that IDs are ints. Internally the EMIS IDs
+    # are SHA512 hashes stored as hex strings which we convert to ints. But we can't
+    # guarantee always to be able to pull this trick.
+    column_specs = {"patient_id": ColumnSpec(int, nullable=False)}
```

Comment:
[Here](https://github.com/opensafely-core/databuilder/blob/0ee050497036e7ed0b242c7b71cde52566881897/databuilder/contracts/types.py#L52) we explicitly allow patient ids to be varchars. Should that affect what you've done here?

Reply:
I'm a bit worried that leaving this as-is will leave us with a potential problem that will be first encountered on a researcher's laptop.

Reply:
I think in that case we should take the varchar ID option out of the contract until we can support it.

Reply:
Yes, I think you're right. I'll create a separate story for that unless you want to do it here.

Reply:
If it's just a case of changing that one contract then I may as well do that here (unless there's further complication I've missed). And then I guess another story for working out how we handle it properly and add it back in?